# 課題2 データの整理

配点
- Q1 4P
- Q2-1 1P
- Q2-2 2P
- Q2-3 1P
- Q2-4 2P


## Q1
### 中央値
中央値はデータの中心傾向を観察するための記述統計の指標です。同じくデータの中心傾向の指標である平均値と比べると、データの外れ値の影響を受けにくという特徴があります。1変数の$n$個のデータを考えた時、中央値を求めるにはまず、データの値の大きさの順に並び替えます。

次に、データの数が奇数個の場合は、大きさの順に並び替えた$(n+1)/2$番目の値、データの数が偶数個の場合は、大きさの順に並び替えた$n/2$番目と$n/2+1$の値の平均の値、をそれぞれデータの中央値とします。

### （標本）分散
分散はデータのばらつきを観察するための記述統計の指標です。個々のデータがデータ全体の平均からどらぐらいばらついているかを表します。以下の通り、個々のデータの値$x_{i}$とデータ全体の平均の値$\overline{x}$の差の2乗和をデータ数$n$で割ったものが分散になります。

$\frac{1}{n}\Sigma^{n}_{i=1}(x_{i}-\overline{x})^{2}$

0から100までの整数を要素とする任意の長さのリストを引数で入力として受け取り、以下を要素とするリストを返す`avg_med_var()`関数を完成させてください。
- 入力リストの全ての要素の平均値
- 入力リストの全ての要素の中央値
- 入力リストの全ての要素の分散

In [19]:
def avg_med_var(nonnegative_int_list):
  list_length = len(nonnegative_int_list)

  average = sum(nonnegative_int_list) / list_length

  median = 0
  nonnegative_int_list.sort()
  mid_index = list_length // 2
  is_even_length = list_length % 2 == 0
  if is_even_length:
    median = (nonnegative_int_list[mid_index - 1] + nonnegative_int_list[mid_index]) / 2
  else:
    median = nonnegative_int_list[mid_index]

  squared_error_sum = 0
  for nonnegative_int in nonnegative_int_list:
    squared_error_sum += (nonnegative_int - average) ** 2
  variance = squared_error_sum / list_length

  return [average, median, variance]

`avg_med_var()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [20]:
print(avg_med_var([1,4,3,2,5,6,8,7,9,10]))
print(avg_med_var([1,4,3,2,5,6,8,7,9,10,11]))

[5.5, 5.5, 8.25]
[6.0, 6, 10.0]


結果はそれぞれ、`[5.5, 5.5, 8.25]`、`[6.0, 6.0, 10.0]`、となります。結果が異なる場合は、作成した`avg_max_min`関数に誤りがないか確認してください。

# Q2
以下のような形式の"user_score.csv"ファイルを読み込み、データフレーム`score`を作成します。
```Python
## user_score.csvファイル
user, kokugo, shakai, sugaku, rika
1, 30, 43, 51, 63
2, 39, 21, 49, 56
...
```

In [11]:
# Colaboratoryでは以下を実行して必要なファイルをダウンロード
!wget https://raw.githubusercontent.com/UTDataMining/2020A/master/ex2/user_score.csv

--2021-10-22 14:20:12--  https://raw.githubusercontent.com/UTDataMining/2020A/master/ex2/user_score.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 2552 (2.5K) [text/plain]
Saving to: ‘user_score.csv’

user_score.csv      100%[===================>]   2.49K  --.-KB/s    in 0s      

2021-10-22 14:20:12 (43.0 MB/s) - ‘user_score.csv’ saved [2552/2552]



In [12]:
import pandas as pd
score = pd.read_csv('user_score.csv')
score.head()

user  kokugo  shakai  sugaku  rika
0     1      30      43      51    63
1     2      39      21      49    56
2     3      29      30      23    57
3     4      95      87      77   100
4     5      70      71      78    67

データフレームの`desribe()`メソッドを使って、各教科ごとの記述統計を確認してください

[pandas.DataFrame.describe](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.describe.html?highlight=describe#pandas.DataFrame.describe)

In [13]:
score.describe()

user      kokugo      shakai      sugaku        rika
count  166.000000  166.000000  166.000000  166.000000  166.000000
mean    83.500000   52.331325   39.602410   45.614458   49.885542
std     48.064193   21.835858   21.456335   24.236127   21.408130
min      1.000000    0.000000    4.000000    0.000000    9.000000
25%     42.250000   35.000000   23.000000   28.000000   31.000000
50%     83.500000   53.000000   35.000000   43.000000   49.000000
75%    124.750000   70.000000   53.750000   63.750000   65.750000
max    166.000000   96.000000   90.000000  100.000000  100.000000

In [22]:
score['sugaku']

0      51
1      49
2      23
3      77
4      78
       ..
161    80
162     2
163    29
164    43
165    89
Name: sugaku, Length: 166, dtype: int64

### Q2-1
データフレーム`score`から指定された科目`subject`の得点が平均点以上の`user`の数を返す`above_mean()`関数を作成してください。

In [23]:
def above_mean(df, subject):
  # subjectは'kokugo','shakai','sugaku','rika'のいずれか
  subject_mean = df[subject].mean()
  return len(df[df[subject] > subject_mean])

`above_mean()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [24]:
above_mean(score, 'kokugo')

84

結果は、`84`、となります。結果が異なる場合は、作成した`above_mean`関数に誤りがないか確認してください。

### Q2-2
ユーザごとの全教科の得点の総和の値を表す'sum'列をデータフレーム`df`に新たに追加したデータフレームを返す`score_sum()`関数を作成してください。

以下のようにして、データフレームの指定した列の行方向の値の和を持つシリーズオブジェクト作ることができます。ユーザごとの全教科の得点の総和の値を表す'sum'列をデータフレーム`df`に新たに追加してください。

```Python
データフレーム[[列1,列2,...,]].sum(axis=1)
```
[pandas.DataFrame.sum](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.sum.html)

In [25]:
def score_sum(df):
  # 教科名は決め打ちでok
  df['sum'] = df[['kokugo', 'shakai', 'sugaku', 'rika']].sum(axis=1)
  return df

`score_sum()`関数が完成したら、以下のセルを実行して動作を確認してください。'sum'列が追加されていることを確認してください。

In [26]:
score = score_sum(score)
score.head()

user  kokugo  shakai  sugaku  rika  sum
0     1      30      43      51    63  187
1     2      39      21      49    56  165
2     3      29      30      23    57  139
3     4      95      87      77   100  359
4     5      70      71      78    67  286

上記で作成したデータフレーム`df`の'sum'列を元に、全教科の合計得点の上位3人の'user'を1位、2位、3位の順に要素として持つリストを返す`score_top3()`関数を作成させてください。なお、合計得点は全て異なるものとし、同一順位はないと仮定してください。

In [58]:
def score_top3(df):
  sorted_df = df.sort_values('sum', ascending=False)
  return [
    sorted_df.iloc[0, 0],
    sorted_df.iloc[1, 0],
    sorted_df.iloc[2, 0],
  ]

`score_top3()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [60]:
score_top3(score)

[39, 4, 50]

結果は、`[39, 4, 50]`、となります。結果が異なる場合は、作成した`score_top3`関数に誤りがないか確認してください。

### Q2-3
データフレーム`score`の'sum'列を対象に、以下を階級幅として各階級の度数を要素として持つリストを返す`score_hist()`関数を完成させてください。
- 0以上、50以下
- 51以上、100以下
- 101以上、150以下
- 151以上、200以下
- 201以上、250以下
- 251以上、300以下
- 301以上、350以下
- 351以上、400以下

In [65]:
def score_hist(df):
  frequency_list = []
  interval_min = 0
  interval_max = 50

  # 0以上、50以下の場合だけ階級幅が異なるため個別に処理
  frequency = len(df[df['sum'].between(interval_min, interval_max)])
  frequency_list.append(frequency)

  for i in range(0, 7):
    interval_min = 51 + i * 50
    interval_max = 100 + i * 50
    frequency = len(df[df['sum'].between(interval_min, interval_max)])
    frequency_list.append(frequency)
 
  return frequency_list

`score_hist()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [66]:
score_hist(score)

[4, 20, 39, 35, 29, 24, 13, 2]

結果は、`[4, 20, 39, 35, 29, 24, 13, 2]`、となります。結果が異なる場合は、作成した`score_hist`関数に誤りがないか確認してください。

### Q2-4
以下のような形式の"user_class.csv"ファイルを読み込み、データフレーム`user_class`を作成します。
```Python
## user_class.csvファイル
'user', 'class'
1, 'C'
2, 'C'
3, 'B'
...
```

In [67]:
# Colaboratoryでは以下を実行して必要なファイルをダウンロード
!wget https://raw.githubusercontent.com/UTDataMining/2020A/master/ex2/user_class.csv

--2021-10-22 16:01:43--  https://raw.githubusercontent.com/UTDataMining/2020A/master/ex2/user_class.csv
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.110.133, 185.199.111.133, 185.199.109.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.110.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 898 [text/plain]
Saving to: ‘user_class.csv’

user_class.csv      100%[===================>]     898  --.-KB/s    in 0s      

2021-10-22 16:01:43 (30.5 MB/s) - ‘user_class.csv’ saved [898/898]



In [68]:
user_class = pd.read_csv('user_class.csv')
user_class.head()

user class
0     1     C
1     2     C
2     3     B
3     4     C
4     5     A

'user'の列の値をキーに、2つのデータフレーム、`user_class`と`score`、を結合して新しいデータフレームを作成した上で、データフレームの`groupby`メソッドを使って、'class'をインデックス、'class'ごとの'sum'の最高値を列として持つシリーズオブジェクトを返す`score_by_class()`関数を完成させてください。

In [79]:
def score_by_class(df1, df2):
  merged_df = pd.merge(df1, df2, on='user')
  return merged_df.groupby('class')['sum'].max()

`score_by_class()`関数が完成したら、以下のセルを実行して動作を確認してください。

In [80]:
score_by_class(score, user_class).to_dict()

{'A': 361, 'B': 349, 'C': 359}

結果は、`{'A': 361, 'B': 349, 'C': 359}`、となります。結果が異なる場合は、作成した`score_by_class`関数に誤りがないか確認してください。

## コードのテスト
以下の実行ボタンを押してから、設問ごとにCheck関数でコードのテストをしてください。

In [81]:
# コードのテストの前にこのセルを実行してください
!pip install prog_edu_assistant_tools
import re
import sys
import jinja2
from IPython.core import display
from google.colab import _message as google_message
from prog_edu_assistant_tools.magics import report, autotest, CaptureOutput
from prog_edu_assistant_tools.check import Check

## Q1

In [82]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q1')

## Q2-1

In [83]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_1')

## Q2-2

In [84]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_2')

## Q2-3

In [85]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_3')

## Q2-4

In [86]:
# Run this cell to check your solution.
# If you get an error 'Check not defined', make sure you have run all preceding
# cells once (Runtime -> Run before)
Check('q2_4')